tools/bpf2c/bpf_code_generator.cpp

```diff
@@ -281,86 +303,109 @@ bpf_code_generator::program_sections()
+bpf_code_generator::generate(const bpf_code_generator::unsafe_string& program_name)
 {
-    current_program = &programs[program_name];
+    bpf_code_generator_program& program = programs[program_name];
 
-    generate_labels();
-    build_function_table();
-    encode_instructions(section_name);
+    program.generate_labels();
+    program.build_function_table();
+    program.encode_instructions(map_definitions);
 }
 
 std::vector<int32_t>
-bpf_code_generator::get_helper_ids()
+bpf_code_generator::get_helper_ids(const bpf_code_generator::unsafe_string& program_name)
 {
+    bpf_code_generator_program& program = programs[program_name];
     std::vector<int32_t> helper_ids;
-    for (const auto& [name, helper] : current_program->helper_functions) {
+    for (const auto& [name, helper] : program.helper_functions) {
         helper_ids.push_back(helper.id);
     }
 
     return helper_ids;
 }
 
 void
-bpf_code_generator::extract_program(const ebpf_api_program_info_t* program_info)
+bpf_code_generator::extract_program(
+    const ebpf_api_program_info_t* program_info, _In_ const ebpf_api_program_info_t* infos)
 {
-    std::vector<ebpf_inst> program{
+    std::vector<ebpf_inst> instructions{
         reinterpret_cast<const ebpf_inst*>(program_info->raw_data),
         reinterpret_cast<const ebpf_inst*>(program_info->raw_data + program_info->raw_data_size)};
 
+    bpf_code_generator_program& program = programs[program_info->program_name];
+
     uint32_t offset = 0;
-    for (const auto& instruction : program) {
-        current_program->output.push_back({instruction, offset++});
+    size_t end_index = instructions.size();
+    for (size_t index = 0; index < end_index; index++) {
+        const auto& instruction = instructions[index];
+
+        // At this point, instructions will contain all subprograms appended.
+        // However, we want only the main program instructions under the main program
+        // function, and subprograms to be under their own function.
+        if (instruction.opcode == INST_OP_CALL && instruction.src == INST_CALL_LOCAL) {
+            size_t callee_index = index + 1 + instruction.imm;
+            if (end_index > callee_index) {
+                end_index = callee_index;
+            }
+            size_t callee_offset = callee_index * sizeof(ebpf_inst);
+
+            // Add subprogram to the list to be parsed.  First we have to find what
+            // the original function name was.  We do this by looking through all other
+            // programs to see which one matches the content at that offset, modulo relocations.
+            for (const ebpf_api_program_info_t* info = infos; info; info = info->next) {
+                if (program_info->raw_data_size < callee_offset + info->raw_data_size) {
+                    continue;
+                }
+                if (memcmp(program_info->raw_data + callee_offset, info->raw_data, info->raw_data_size) == 0) {
```

Comment:
I don't understand what this is doing. How is the code determining the beginning and end of the sub-program?

Reply:
Added a comment explaining.